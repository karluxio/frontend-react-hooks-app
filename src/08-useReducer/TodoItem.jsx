export const TodoItem = ({ todo, onDeleteTodo, onToggleTodo }) => {
  return (
    <li className="d-flex justify-content-between list-group-item">

      <span
        onClick={() => onToggleTodo(todo.id)}
        className={`align-self-center ${(todo.done) ? 'text-decoration-line-through' : 'text-decoration-none'}`}
      >
        {todo.description}
      </span>

      <button
        onClick={() => onDeleteTodo(todo.id)}
        className="btn btn-danger">
        delete
      </button>

    </li >
  )
}