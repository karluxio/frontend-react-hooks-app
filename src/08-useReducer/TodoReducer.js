
export const initialState = []

export const todoReducer = (state = initialState, { type, payload }) => {
  switch (type) {

    case '[TODO] add':
      return [...state, { ...payload }]

    case '[TODO] toggle':
      return [...state].map(todo => {
        if (todo.id === payload) {
          return {
            ...todo,
            done: !todo.done
          }
        }
        return todo
      })

    case '[TODO] delete':
      return [...state].filter(todo => todo.id !== payload)

    default:
      return state
  }
}
