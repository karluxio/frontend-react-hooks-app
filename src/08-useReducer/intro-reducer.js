export const initialState = [
  {
    id: 1,
    todo: 'recollect the soul gem',
    done: false
  }
]

export const todosReducer = (state = initialState, action = {}) => {

  if (action.type === "[TODO] add todo") {
    return [...state, action.payload]
  }

  return state
}

let todos = todosReducer()

const newTodo = {
  id: 2,
  todo: "recollect the power gem",
  done: true
}

const addTodoAction = {
  type: '[TODO] add todo',
  payload: newTodo
}

todos = todosReducer(todos, addTodoAction)

console.log({ todos });
