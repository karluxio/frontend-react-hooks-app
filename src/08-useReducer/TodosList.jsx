import { TodoItem } from "./TodoItem"

export const TodosList = ({ todos = [], onDeleteTodo, onToggleTodo }) => {
  return (
    <ul className="list-group">
      {
        todos.length > 0 && todos.map(todo => (
          <TodoItem key={todo.id} todo={todo} onDeleteTodo={onDeleteTodo} onToggleTodo={onToggleTodo} />
        ))
      }
    </ul>
  )
}