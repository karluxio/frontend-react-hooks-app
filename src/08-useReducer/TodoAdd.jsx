import { useForm } from "../hooks/useForm"

export const TodoAdd = ({ onNewTodo }) => {

  const { description, onInputChange, onResetForm } = useForm({ description: '' })

  const onSubmit = (e) => {
    e.preventDefault()

    if (description <= 1) return
    onNewTodo({
      id: new Date().getTime(),
      description,
      done: false
    })

    onResetForm()
  }

  return (
    <>
      <h4>Add todo</h4>
      <hr />
      <form onSubmit={onSubmit}>
        <input
          className="form-control"
          name="description"
          onChange={onInputChange}
          placeholder="cleanup my house"
          type="text"
          value={description}
        />

        <button type="submit" className="btn btn-outline-primary mt-2">Add</button>
      </form>
    </>
  )
}