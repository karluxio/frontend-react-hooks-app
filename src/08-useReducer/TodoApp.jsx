import { useReducer, useEffect } from "react"
import { todosReducer, initialState } from "./TodoReducer"
import { TodosList } from "./TodosList"
import { TodoAdd } from "./TodoAdd"

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || []
}

export const TodoApp = () => {

  const [todos, dispatch] = useReducer(todosReducer, initialState, init)

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos))
  }, [todos])


  const handleNewTodo = (todo) => {
    dispatch({ type: '[TODO] add', payload: todo })
  }

  const handleDeleteTodo = (id) => {
    dispatch({ type: '[TODO] delete', payload: id })
  }

  const handleToggleTodo = (id) => {
    dispatch({ type: '[TODO] toggle', payload: id })
  }

  return (
    <>

      <h1>TodoApp: {todos.length}, <small>pending: 4</small></h1>
      <hr />

      <div className="row">

        <div className="col-7">
          <TodosList todos={todos} onDeleteTodo={handleDeleteTodo} onToggleTodo={handleToggleTodo} />
        </div>

        <div className="col-5">
          <TodoAdd onNewTodo={handleNewTodo} />
        </div>

      </div>

    </>
  )
}