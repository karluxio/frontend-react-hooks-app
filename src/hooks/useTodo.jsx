import { useEffect, useReducer } from "react"
import { todosReducer } from "../08-useReducer/TodoReducer"

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || []
}

export const useTodo = () => {

  const [todos, dispatch] = useReducer(todosReducer, [], init)

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos))
  }, [todos])


  const handleNewTodo = (todo) => {
    dispatch({ type: '[TODO] add', payload: todo })
  }

  const handleDeleteTodo = (id) => {
    dispatch({ type: '[TODO] delete', payload: id })
  }

  const handleToggleTodo = (id) => {
    dispatch({ type: '[TODO] toggle', payload: id })
  }

  const todosCount = () => todos.length

  const pendingTodosCount = () => todos.filter(todo => todo.done === false).length

  return {
    todos,
    handleNewTodo,
    handleDeleteTodo,
    handleToggleTodo,
    todosCount,
    pendingTodosCount
  }
}