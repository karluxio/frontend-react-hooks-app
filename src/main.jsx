// import React from 'react'
import ReactDOM from 'react-dom/client'

// import { HooksApp } from './HooksApp'
// import { CounterApp } from './01-useState/CounterApp'
// import { CounterWithCustomHook } from './01-useState/CounterWithCustomHook'
// import { SimpleForm } from './02-useEffect/SimpleForm'
// import { FormWithCustomForm } from './02-useEffect/FormWithCustomHook'
// import { MultiplesCustomHooks } from './03-examples/MultiplesCustomHooks'
// import { FocusScreen } from './04-useRef/FocusScreen'
// import { Layout } from './05-useLayoutEffect/Layout'
// import { Memorize } from './06-memos/Memorize'
// import { MemoHook } from './06-memos/MemoHook'
// import { CallbackHook } from './06-memos/CallbackHook'
// import { Padre } from './07-tarea-memo/Padre'

// import './08-useReducer/intro-reducer'

import './index.css'
import { TodoApp } from './08-useReducer/TodoApp'

ReactDOM.createRoot(document.getElementById('root')).render(
  //   // <React.StrictMode>
  <>
    {/* <HooksApp /> */}
    {/* <CounterApp />*/}
    {/* <CounterWithCustomHook /> */}
    {/* <SimpleForm /> */}
    {/* <FormWithCustomForm /> */}
    {/* <MultiplesCustomHooks /> */}
    {/* <FocusScreen /> */}
    {/* <Layout /> */}
    {/* <Memorize /> */}
    {/* <MemoHook /> */}
    {/* <CallbackHook /> */}
    {/* <Padre /> */}
    <TodoApp />
  </>


  // </React.StrictMode>,
)
