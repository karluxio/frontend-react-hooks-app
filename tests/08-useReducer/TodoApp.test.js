import { render, screen } from '@testing-library/react'
import { useTodo } from '../../src/hooks'
import { TodoApp } from '../../src/08-useReducer/TodoApp'

jest.mock('../../src/hooks/useTodo.js')

const todos = [
  {
    id: 1,
    description: 'some description',
    done: false
  }
]

describe("TodoApp", () => {

  beforeEach(() => {
    jest.clearAllMocks()
  })

  const handleNewTodoMock = jest.fn()
  const handleDeleteTodoMock = jest.fn()
  const handleToggleTodoMock = jest.fn()
  const todosCountMock = jest.fn(() => todos.length)
  const pendingTodosCountMock = jest.fn(() => todos.filter(todo => todo.done !== true).length)

  useTodo.mockReturnValue({
    todos,
    handleNewTodoMock,
    handleDeleteTodoMock,
    handleToggleTodoMock,
    todosCount: todosCountMock,
    pendingTodosCount: pendingTodosCountMock
  })

  it("should match against snapshot", () => {
    const { container } = render(<TodoApp />)
    expect(container).toMatchSnapshot()
  })

  it("should render amount and pending todos in title", () => {
    render(<TodoApp />)

    expect(screen.getByRole("heading", { level: 1 }).innerHTML).toContain(`TodoApp: ${todos.length}`)
    expect(screen.getByRole("heading", { level: 1 }).innerHTML).toContain(
      `pending: ${todos.filter(todo => todo.done !== true).length}`
    )
    // screen.debug()
  })

  it("should render a list of todos with length 1", () => {
    render(<TodoApp />)
    expect(screen.getAllByRole('listitem')).toHaveLength(1)
  })

  it("should render an input", () => {
    render(<TodoApp />)
    expect(screen.getByRole("textbox")).toBeTruthy()
  })
})