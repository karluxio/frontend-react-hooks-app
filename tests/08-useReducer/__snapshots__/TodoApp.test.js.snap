// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp should match against snapshot 1`] = `
<div>
  <h1>
    TodoApp: 
    1
    , 
    <small>
      pending: 
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="d-flex justify-content-between list-group-item"
        >
          <span
            class="align-self-center text-decoration-none"
          >
            some description
          </span>
          <button
            class="btn btn-danger"
          >
            delete
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Add todo
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="description"
          placeholder="cleanup my house"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</div>
`;
